contracts/services/reverse-splitter/src/contract.rs

```diff
@@ -0,0 +1,299 @@
+
+            if let Some(RatioConfig::DynamicRatio {
+                contract_addr,
+                params,
+            }) = split.ratio()
+            {
+                let key = dyn_ratio_key(denom, contract_addr, params);
+                if let Entry::Vacant(e) = dynamic_ratios.entry(key) {
+                    let ratio = query_dynamic_ratio(querier, contract_addr, params, denom)?;
+                    e.insert(ratio);
+                }
+            }
+        }
+
+        if denom_amount_count == cfg.splits().len() {
+            // If all splits have an amount (and we have checked the balances),
+            // we can return the amounts as-is.
+            return cfg
+                .splits()
+                .iter()
+                .map(|split| Ok((split.amount().unwrap(), split.denom(), split.account())))
+                .collect::<Result<Vec<_>, _>>();
+        }
+
+        // If not all splits have an amount, we need to compute the amounts based on the ratios
+
+        // Prepare transfer messages for each split config
+        let amounts_in_base_denom = cfg
+            .splits()
+            .iter()
+            .map(|split| {
+                // Lookup account/denom balance
+                let account = split.account();
+                let denom = split.denom();
+                let balance = account_balances.get(&account_key(account, denom)).unwrap();
+                split
+                    .amount()
+                    .map(|amount| {
+                        assert_eq!(denom, cfg.base_denom());
+                        Ok((amount, Decimal::one(), &None, denom, account))
+                    })
+                    .or_else(|| {
+                        split.ratio().as_ref().map(|ratio_config| {
+                            let ratio = match ratio_config {
+                                RatioConfig::FixedRatio(ratio) => ratio,
+                                RatioConfig::DynamicRatio {
+                                    contract_addr,
+                                    params,
+                                } => dynamic_ratios
+                                    .get(&dyn_ratio_key(split.denom(), contract_addr, params))
+                                    .unwrap(),
+                            };
+                            let mut amount =
+                                balance.multiply_ratio(ratio.numerator(), ratio.denominator());
+                            let factor = split.factor();
+                            if let Some(factor) = factor {
+                                amount = amount
+                                    .checked_div((*factor as u128).into())
+                                    .map_err(|err| ServiceError::Std(err.into()))?;
+                            }
+                            Ok((amount, *ratio, factor, denom, account))
+                        })
+                    })
+                    .expect("Split config must have either an amount or a ratio")
```

Comment:
Nit: This is unreachable because we validated it right? Might add it here in a comment or something

Reply:
Yes, it is an invariant .. could have done an unwrap() instead.